src/main/java/de/tk/opensource/secon/KeyStoreDirectory.java

```diff
@@ -35,41 +35,64 @@
  */
 final class KeyStoreDirectory implements Directory {
 
-    private final KeyStore ks;
+	private final KeyStore ks;
+	
+	private final LRUCache<X509Certificate, X509Certificate> issuerCache;
+	
+	KeyStoreDirectory(final KeyStore ks) {
+		this(ks, new LRUCache<X509Certificate, X509Certificate>(50));
+	}
+	
+	KeyStoreDirectory(final KeyStore ks, LRUCache<X509Certificate, X509Certificate> cache) {
+		this.ks = ks;
+		this.issuerCache = cache;
+	}
 
-    KeyStoreDirectory(final KeyStore ks) {
-        this.ks = ks;
-    }
+	@Override
+	public final Optional<X509Certificate> certificate(X509CertSelector selector) throws KeyStoreException {
+		return certificates(selector).findFirst();
+	}
 
-    @Override
-    public final Optional<X509Certificate> certificate(X509CertSelector selector) throws KeyStoreException {
-        return Collections
-                .list(ks.aliases())
-                .stream()
-                .flatMap(this::certificateStream)
-                .filter(selector::match)
-                .findFirst();
-    }
+	private final Stream<X509Certificate> certificates(X509CertSelector selector) throws KeyStoreException {
+		return Collections.list(ks.aliases()).stream().flatMap(this::certificateStream).filter(selector::match);
+	}
 
-    @Override
+	@Override
 	public Optional<X509Certificate> issuer(X509Certificate cert) throws Exception {
-    	final X509CertSelector selector = new X509CertSelector();
-    	selector.setSubject(cert.getIssuerX500Principal());
-    	
-		return certificate(selector);
+		if(issuerCache.containsKey(cert)) {
+			return issuerCache.get(cert);
+		}
+		
+		final X509CertSelector selector = new X509CertSelector();
+		selector.setSubject(cert.getIssuerX500Principal());
+
+		// find first matching issuer since principal might not be unique
+		Optional<X509Certificate> issuer = certificates(selector).filter(p -> {
```

Comment:
correct issuer can only be identified by verification of signature. E.g. ITSG reuses the same principal name when renewing issuer certs. Thus we can have different issuer certificates for the same issuer principal.